### PR TITLE
Fix #18648: Fix `yii\web\Request` to properly handle HTTP Basic Auth …

### LINKED ISSUE #18648: Http Basic Auth doesn't work
### What steps will reproduce the problem?
Http Basic Auth doesn't work if the server uses a special configuration where PHP is not loaded as module...

References
#13654
#18315

### Additional info
In the yii\web\Request::getAuthCredentials()
```
/*
 * Apache with php-cgi does not pass HTTP Basic authentication to PHP by default.
 * To make it work, add the following line to to your .htaccess file:
 *
 * RewriteRule .* - [E=HTTP_AUTHORIZATION:%{HTTP:Authorization}]
 */
$auth_token = $this->getHeaders()->get('HTTP_AUTHORIZATION') ?: $this->getHeaders()->get('REDIRECT_HTTP_AUTHORIZATION');
```
**The problem is that HTTP_AUTHORIZATION and REDIRECT_HTTP_AUTHORIZATION are not HEADERS, but environment variables!** 

In the yii\web\Request::getHeaders() HTTP headers parsed by
`getallheaders()`
or
`http_get_request_headers()`
or
custom expression `(strncmp($name, 'HTTP_', 5) === 0)` - will not return REDIRECT_HTTP_AUTHORIZATION as a header.

In the tests/framework/web/RequestTest.php
```
testHttpAuthCredentialsFromHttpAuthorizationHeader()
$request->getHeaders()->set('HTTP_AUTHORIZATION', 'Basic ' . $secret);
$request->getHeaders()->set('REDIRECT_HTTP_AUTHORIZATION', 'Basic ' . $secret);
```
It's not correct to set headers! You must set ENV variables!

### Solution
1. Get auth data from 'Authorization' header not from env vars, replace
```
$auth_token = $this->getHeaders()->get('HTTP_AUTHORIZATION') ?: $this->getHeaders()->get('REDIRECT_HTTP_AUTHORIZATION');
```
with
`$auth_token = $this->getHeaders()->get('Authorization');`

2. Update custom headers parser and add expression for 'REDIRECT_HTTP_' prefix:
`(strncmp($name, 'HTTP_', 5) === 0 || strncmp($name, 'REDIRECT_HTTP_', 14) === 0)`

### PR DESCRIPTION
**Pull request by olegbaturin**
…headers

| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ✔️
| Tests pass?   | ✔️
| Fixed issues  | #18648

### CODE REVIEW

File: framework/web/Request.php [hunk 1]
```diff
@@ -375,7 +375,11 @@ public function getHeaders()
                     if (strncmp($name, 'HTTP_', 5) === 0) {
                         $name = str_replace(' ', '-', ucwords(strtolower(str_replace('_', ' ', substr($name, 5)))));
                         $this->_headers->add($name, $value);
+                    } elseif (strncmp($name, 'REDIRECT_HTTP_', 14) === 0) {
```
Comment (bizley) [inline framework/web/Request.php:378]: Do we expect something else here than `REDIRECT_HTTP_AUTHORIZATION`? If not - there is no reason for the next line to be so sophisticated.

File: framework/web/Request.php [hunk 2]
```diff
@@ -375,7 +375,11 @@ public function getHeaders()
...
+                        $name = str_replace(' ', '-', ucwords(strtolower(str_replace('_', ' ', substr($name, 14)))));
+                        $this->_headers->add($name, $value);
                     }
+
```
Comment (bizley) [inline framework/web/Request.php:382]: Please remove empty line

**Review by bizley — CHANGES_REQUESTED** [review]

File: framework/web/Request.php (hunk 1 above)
Comment (olegbaturin) [reply framework/web/Request.php:378]: Sounds like a prequel to the next bug. Request::getHeaders() is a global headers parser method, so we can expect any header here that were set in RewriteRule as environment variable.

Comment (bizley) [reply framework/web/Request.php:378]: Not sure if I understand what you said but I think you mean that there can be something like `REDIRECT_HTTP_WHATEVER` and we should add `Whatever` then to headers. If so, then ok, I got it now.

Comment (olegbaturin) [reply framework/web/Request.php:378]: We can parse only "REDIRECT_" prefix for any headers, so the rules for Authorization header will become
```
RewriteRule .* - [E=AUTHORIZATION:%{HTTP:Authorization}]
SetEnvIf Authorization .+ AUTHORIZATION=$0
```
and env var name will become REDIRECT_AUTHORIZATION.

As I understand HTTP_AUTHORIZATION was used in the Rewrite rule later, before Apache start prefixes env vars with "REDIRECT_".

Comment (olegbaturin) [reply framework/web/Request.php:378]: > Not sure if I understand what you said but I think you mean that there can be something like `REDIRECT_HTTP_WHATEVER` and we should add `Whatever` then to headers. If so, then ok, I got it now.

In this rule `SetEnvIf Authorization .+ AUTHORIZATION=$0` you can set any env var name for any http header.

**Comment by olegbaturin** [comment]
I make a small update of the headers parser method. Header prefixes of headers in the $_SERVER variable now in $headerPrefixes attribute.

File: framework/web/Request.php [hunk 3]
```diff
@@ -261,6 +261,14 @@ class Request extends \yii\base\Request
         'Front-End-Https' => ['on'], // Microsoft
     ];
 
+    /**
+     * @var array list of the key prefixes used to obtain HTTP headers from $_SERVER variable.
+     */
+    public $headerPrefixes = [
```
Comment (bizley) [inline framework/web/Request.php:267]: I'm not sure if we want to expose that. We were quite fine with one prefix so far, and the second one looks like an exception from the rule only.

**Comment by olegbaturin** [comment]
I can make a new clear pull request with one commit.

**Comment by bizley** [comment]
No need, this is fine. Some build are failing though, that needs to be addressed first.

**Comment by bizley** [comment]
@olegbaturin any chance for fixing the bugs?

**Comment by olegbaturin** [comment]
What bugs?

**Comment by bizley** [comment]
At the bottom of this page you can see the list of all the builds that are automatically run by Github actions. 10 of them failed.

**Comment by olegbaturin** [comment]
Yes, I'll fix them all.

**Review by bizley — APPROVED** [review]

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Thank you!

**Pull request merged**